qiskit/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -0,0 +1,219 @@
+        self._qubits = qubits
+        self._spacing = spacing
+        self._skip_reset_qubits = skip_reset_qubits
+
+    def run(self, dag):
+        """Run the DynamicalDecoupling pass on dag.
+
+        Args:
+            dag (DAGCircuit): a scheduled DAG.
+
+        Returns:
+            DAGCircuit: equivalent circuit with delays interrupted by DD,
+                where possible.
+
+        Raises:
+            TranspilerError: if the circuit is not mapped on physical qubits.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get("q", None) is None:
+            raise TranspilerError("DD runs on physical circuits only.")
+
+        if dag.duration is None:
+            raise TranspilerError("DD runs after circuit is scheduled.")
+
+        num_pulses = len(self._dd_sequence)
+        sequence_gphase = 0
+        if num_pulses != 1:
+            if num_pulses % 2 != 0:
+                raise TranspilerError("DD sequence must contain an even number of gates (or 1).")
+            noop = np.eye(2)
+            for gate in self._dd_sequence:
+                noop = noop.dot(gate.to_matrix())
+            if not matrix_equal(noop, IGate().to_matrix(), ignore_phase=True):
+                raise TranspilerError("The DD sequence does not make an identity operation.")
+            sequence_gphase = np.angle(noop[0][0])
+
+        if self._qubits is None:
+            self._qubits = range(dag.num_qubits())
+
+        if self._spacing:
+            if sum(self._spacing) != 1 or any(a < 0 for a in self._spacing):
+                pass
+                raise TranspilerError(
+                    "The spacings must be given in terms of fractions "
+                    "of the slack period and sum to 1."
+                )
+        else:  # default to balanced spacing
+            mid = 1 / num_pulses
+            end = mid / 2
+            self._spacing = [end] + [mid] * (num_pulses - 1) + [end]
+
+        new_dag = dag._copy_circuit_metadata()
+
+        for nd in dag.topological_op_nodes():
+            if not isinstance(nd.op, Delay):
+                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                continue
+
+            dag_qubit = nd.qargs[0]
+            physical_qubit = new_dag.qubits.index(dag_qubit)
+            if physical_qubit not in self._qubits:  # skip unwanted qubits
+                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                continue
+
+            pred = next(dag.predecessors(nd))
+            succ = next(dag.successors(nd))
+            if self._skip_reset_qubits:  # discount initial delays
+                if pred.type == "in" or isinstance(pred, Reset):
+                    new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                    continue
+
+            dd_sequence_duration = 0
+            for gate in self._dd_sequence:
+                gate.duration = self._durations.get(gate, physical_qubit)
+                dd_sequence_duration += gate.duration
```

Comment:
Would it make sense to do this outside the for loop, since it should be static based on the configured dd sequence we don't need to recompute it for every delay.

Reply:
Lauren mentioned the same thing above. The duration is different for every qubit, but yes same across differernt delays of the same qubit. I did this in https://github.com/Qiskit/qiskit-terra/pull/6619/commits/fa9da7882c93876a4217eaae15eb4a974ba28093